porcupine/plugins/anchors.py

```diff
@@ -0,0 +1,143 @@
+"""
+Allow for easy navigation in your files by toggling on/off anchorpoints,
+and jumpbing back and forth between them
+"""
+from __future__ import annotations
+
+import time
+import tkinter
+
+from porcupine import get_tab_manager, settings, tabs
+from porcupine.plugins.linenumbers import LineNumbers
+
+# Dependent on code from linenumbers.py
+setup_after = ["linenumbers"]
+
+# TODO: Add checkbox to settings window to allow/disallow cycling from last anchor to first anchor with "jump_to_next"
+
+
+class AnchorManager:
+    def __init__(self, tab_textwidget: tkinter.Text, linenumbers: LineNumbers) -> None:
+        self.tab_textwidget = tab_textwidget
+        self.linenumbers = linenumbers
+
+        self.anchor_symbol = "¶"
+
+        # tkinter have default marks such as "insert", "current", "tkinter::anchor1"
+        self.custom_anchor_prefix = "anchor_"
+
+        linenumbers.bind("<<Updated>>", self.update_linenumbers, add=True)
+
+    def _get_anchors(self) -> list[str]:
+        return [
+            anchor
+            for anchor in self.tab_textwidget.mark_names()
+            if anchor.startswith(self.custom_anchor_prefix)
+        ]
+
+    def _get_cursor_index(self) -> str:
+        return self.tab_textwidget.index("insert linestart")
+
+    def toggle_on_off(self, event: tkinter.Event[tkinter.Misc]) -> None:
+
+        self.prevent_duplicate_anchors()
+
+        cursor_index = self._get_cursor_index()
+        for anchor in self._get_anchors():
+            if self.tab_textwidget.index(anchor) == cursor_index:
+                self.tab_textwidget.mark_unset(anchor)
+                self.linenumbers.do_update()
+                return
+        self.tab_textwidget.mark_set(
+            self.custom_anchor_prefix + str(time.time()), "insert linestart"
+        )
+
+        self.linenumbers.do_update()
+
+    def jump_to_next(self, event: tkinter.Event[tkinter.Misc]) -> str:
+        cursor_row = self._get_cursor_index().split(".")[0]
+        anchor_list = self._get_anchors()
+        anchor_rows = [
+            int(self.tab_textwidget.index(anchorpoint).split(".")[0]) for anchorpoint in anchor_list
+        ]
+
+        rows_after_cursor = [n for n in anchor_rows if n > int(cursor_row)]
+        if rows_after_cursor:
+            next_anchor_row = min(rows_after_cursor)
+            self.tab_textwidget.mark_set("insert", f"{next_anchor_row}.0")
+            self.tab_textwidget.see("insert")
+        elif not rows_after_cursor and len(anchor_list) >= 2 and settings.get("anchors", bool):
```

Comment:
Nit: `rows_after_cursor` will always be False when this condition is considered, because if it isn't, then we choose the first part.

```suggestion
        elif len(anchor_list) >= 2 and settings.get("anchors", bool):
```

Reply:
Genius.